In [2]:
import pandas as pd

## Задание 1

Для датафрейма log из материалов занятия создайте столбец source_type по следующим правилам:

- если источник traffic_source равен yandex или google, то в source_type ставится organic
- для источников paid и email из России - ставим ad
- для источников paid и email не из России - ставим other
- все остальные варианты берем из traffic_source без изменений

In [3]:
df_log = pd.read_csv('visit_log.csv', sep=';')
df_log.head()

timestamp    visit_id                                 url  region  \
0  1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3  Russia   
1  1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a  Russia   
2  1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15  Russia   
3  1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15  Russia   
4  1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15  Russia   

      user_id traffic_source  
0  b1613cc09f         yandex  
1  4c3ec14bee         direct  
2  a8c40697fb         yandex  
3  521ac1d6a0         yandex  
4  d7323c571c         yandex

In [11]:
df_log.loc[ (df_log.traffic_source == 'yandex') | (df_log.traffic_source == 'google'), 'source_type'] = 'organic'
df_log.loc[ ( (df_log.traffic_source == 'email') | (df_log.traffic_source == 'paid') ) 
           & (df_log.region == 'Russia'), 'source_type'] = 'ad'
df_log.loc[ (df_log.traffic_source.isin(['paid', 'email']))
           & (df_log.region != 'Russia'), 'source_type'] = 'other'
df_log.sample(10)

timestamp    visit_id                                 url   region  \
3249   1550000225  93097c3618  https://host.ru/8e88d4703848cc0ec4   Russia   
12096  1550053003  17919b5667  https://host.ru/ac2d98a5a63e7c6e28  Belarus   
8620   1550032285  9678b994ae  https://host.ru/108ce4b365afb7b88e   Russia   
13170  1550059163  a917e0da47  https://host.ru/da83c74e35e7806fe0  Belarus   
15667  1550074374  ffdf07332b  https://host.ru/186c4d6b47dbde0e74  Belarus   
1877   1549991973  0ad54315bb  https://host.ru/23c64743ad842b9311    China   
10879  1550045622  428a05e558  https://host.ru/ee7c4e0674a284ee3f   Russia   
610    1549984337  410eff9993  https://host.ru/8880fd00cf23066541   Russia   
17894  1550087979  0c8739aa6f  https://host.ru/1a0fe145099b114845  Ukraine   
11715  1550050732  cb1340343c  https://host.ru/23c64743ad842b9311  Ukraine   

          user_id traffic_source source_type  
3249   92daf41708          email          ad  
12096  edba6d294f         yandex     organic  
8620   da3d0467e5         direct         NaN  
13170  a87f645065         direct         NaN  
15667  5873e24b5b           paid       other  
1877   0e1752c18e         yandex     organic  
10879  7cd406b116           paid          ad  
610    72f6ac0129         google     organic  
17894  a4d2680e71          email       other  
11715  eec374ffcc           paid       other

In [13]:
df_log.loc[pd.isna(df_log.source_type), 'source_type'] = df_log.traffic_source
df_log.sample(10)

timestamp    visit_id                                 url   region  \
7181   1550023779  fc563e693b  https://host.ru/3004a8273caeef2867   Russia   
3937   1550004358  6aa580d8f5  https://host.ru/060d9bdbcba66b8bb3   Russia   
16804  1550081247  04bb33b2d0  https://host.ru/108ce4b365afb7b88e   Russia   
3804   1550003553  659d4349dc  https://host.ru/05e41944dd85a900a6  Belarus   
1039   1549986882  b7486677b3  https://host.ru/53b1649b8344e663e9   Russia   
17599  1550086088  757539f9dd  https://host.ru/53b1649b8344e663e9    China   
8542   1550031785  b55f817f14  https://host.ru/3a80d6ca023aed7752    India   
15393  1550072828  3a96a2aa0c  https://host.ru/3004a8273caeef2867  Ukraine   
17358  1550084656  938f9a5407  https://host.ru/186c4d6b47dbde0e74   Russia   
11238  1550047881  4582bd24ce  https://host.ru/8880fd00cf23066541  Belarus   

          user_id traffic_source source_type  
7181   5682b76a24         direct      direct  
3937   e9c3bbc30a         direct      direct  
16804  ce6c812701         google     organic  
3804   508693b1ab         direct      direct  
1039   2cfec347df          email          ad  
17599  373b29c11b         direct      direct  
8542   ae98be30ea         direct      direct  
15393  7432f4e5a9         yandex     organic  
17358  13fcf56cd7         direct      direct  
11238  fb9f3f40d8          email       other

## Задание 2

В файле URLs.txt содержатся url страниц новостного сайта. Вам необходимо отфильтровать его по адресам страниц с текстами новостей. Известно, что шаблон страницы новостей имеет внутри url следующую конструкцию: /, затем 8 цифр, затем дефис. Выполните следующие действия:

1. Прочитайте содержимое файла в датафрейм
2. Отфильтруйте страницы с текстом новостей, используя метод str.contains и регулярное выражение в соответствии с заданным шаблоном

In [17]:
import re

In [14]:
df_urls = pd.read_csv('URLs.txt')
df_urls.head()

url
0                                            /world/
1                                           /latest/
2                                      /?updated=top
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...

In [20]:
df_urls.loc[df_urls.url.str.contains(r'[/]\d{8}[-]')].sample(10)

url
25  /politics/36150505-stala-izvestna-prichina-sme...
72  /kinomusic/36016432-igorya-krutogo-prooperirov...
63  /business/36184872-udar-po-schetchiku-rossiyan...
83  /world/36215106-lider-ig-obratilsya-k-boevikam...
84  /articles/36066499-iz-za-zvaniy-anny-netrebko-...
81  /articles/36093326-shturmovat-donetsk-i-lugans...
68  /video/36202662-nakazanie-dlya-lukashenko-ross...
3   /politics/36188461-s-marta-zhizn-rossiyan-susc...
23  /politics/36115220-mid-vystupil-s-preduprezhde...
14  /politics/35638742-nesolidno-mariya-zaharova-s...

## Задание 3

Используйте файл с оценками фильмов ml-latest-small/ratings.csv. Посчитайте среднее время жизни пользователей, которые выставили более 100 оценок. Под временем жизни понимается разница между максимальным и минимальным значением столбца timestamp для данного значения userId.

In [33]:
df_user_lifetime = pd.read_csv('ml-latest-small/ratings.csv')
df_user_lifetime.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [43]:
user_max = df_user_lifetime[['userId', 'timestamp']].groupby('userId').max().reset_index()
user_max.head()

userId   timestamp
0       1  1260759205
1       2   835356246
2       3  1298932787
3       4   949982274
4       5  1163375145

In [44]:
user_min = df_user_lifetime[['userId', 'timestamp']].groupby('userId').min().reset_index()
user_min.head()

userId   timestamp
0       1  1260759108
1       2   835355395
2       3  1298861589
3       4   949778714
4       5  1163373044

In [53]:
# количество оценок пользователей
user_count = df_user_lifetime[['userId', 'timestamp']].groupby('userId').count().reset_index()
user_count.head()

userId  timestamp
0       1         20
1       2         76
2       3         51
3       4        204
4       5        100

In [54]:
# фильтр пользователей с оценкаим больше 100
user100 = user_count[user_count.timestamp > 100]
user100.head()

userId  timestamp
3        4        204
7        8        116
14      15       1700
16      17        363
18      19        423

In [67]:
# теперь начинаем все это соединять
total = user100.merge(user_max, on='userId', how='left').merge(user_min, on='userId', how='left')
total.head(10)

userId  timestamp_x  timestamp_y   timestamp
0       4          204    949982274   949778714
1       8          116   1154474527  1154389340
2      15         1700   1469330735   997937239
3      17          363   1127476640  1127468587
4      19          423    855195373   855190091
5      21          162    854522908   853157476
6      22          220   1131753381  1131661890
7      23          726   1166728253  1148386124
8      26          172   1371811577  1351544316
9      30         1011   1060795346   944943070

In [68]:
total.rename(columns = {'timestamp_x': 'count', 'timestamp_y': 'time_out', 'timestamp': 'time_in'}, inplace=True)
total.head()

userId  count    time_out     time_in
0       4    204   949982274   949778714
1       8    116  1154474527  1154389340
2      15   1700  1469330735   997937239
3      17    363  1127476640  1127468587
4      19    423   855195373   855190091

In [76]:
# время жизни в днях каждого пользователя (86400 - кол-во секунд в сутках)
total['lifetime'] = round((total.time_out - total.time_in) / 86400, 2)
total.head(10)

userId  count    time_out     time_in  lifetime
0       4    204   949982274   949778714      2.36
1       8    116  1154474527  1154389340      0.99
2      15   1700  1469330735   997937239   5455.94
3      17    363  1127476640  1127468587      0.09
4      19    423   855195373   855190091      0.06
5      21    162   854522908   853157476     15.80
6      22    220  1131753381  1131661890      1.06
7      23    726  1166728253  1148386124    212.29
8      26    172  1371811577  1351544316    234.57
9      30   1011  1060795346   944943070   1340.88

In [78]:
# Ну и ответ на поставленный вопрос - среднее время жизни пользователей со 100+ оценками
round(total['lifetime'].mean(), 2)

463.89

### Ответ 463.89

## Задание 4

Дана статистика услуг перевозок клиентов компании по типам (см. файл с кодом занятия). Необходимо сформировать две таблицы:
- таблицу с тремя типами выручки для каждого client_id без указания адреса клиента
- аналогичную таблицу по типам выручки с указанием адреса клиента

Обратите внимание, что в процессе объединения таблиц данные не должны теряться.

### К домашнему заданию
Дана статистика услуг перевозок клиентов компании по типам:
- rzd - железнодорожные перевозки
- auto - автомобильные перевозки
- air - воздушные перевозки
- client_base - адреса клиентов

In [3]:
rzd = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115],
        'rzd_revenue': [1093, 2810, 10283, 5774, 981]
    }
)
rzd

client_id  rzd_revenue
0        111         1093
1        112         2810
2        113        10283
3        114         5774
4        115          981

In [4]:
auto = pd.DataFrame(
    {
        'client_id': [113, 114, 115, 116, 117],
        'auto_revenue': [57483, 83, 912, 4834, 98]
    }
)
auto

client_id  auto_revenue
0        113         57483
1        114            83
2        115           912
3        116          4834
4        117            98

In [5]:
air = pd.DataFrame(
    {
        'client_id': [115, 116, 117, 118],
        'air_revenue': [81, 4, 13, 173]
    }
)
air

client_id  air_revenue
0        115           81
1        116            4
2        117           13
3        118          173

In [6]:
client_base = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115, 116, 117, 118],
        'address': ['Комсомольская 4', 'Энтузиастов 8а', 'Левобережная 1а', 'Мира 14', 'ЗЖБИиДК 1', 
                    'Строителей 18', 'Панфиловская 33', 'Мастеркова 4']
    }
)
client_base

client_id          address
0        111  Комсомольская 4
1        112   Энтузиастов 8а
2        113  Левобережная 1а
3        114          Мира 14
4        115        ЗЖБИиДК 1
5        116    Строителей 18
6        117  Панфиловская 33
7        118     Мастеркова 4

In [9]:
df_revenue = rzd.merge(auto, on='client_id',  how='outer').merge(air, on='client_id', how='outer')
df_revenue.fillna(0.0, inplace=True)
df_revenue

client_id  rzd_revenue  auto_revenue  air_revenue
0        111       1093.0           0.0          0.0
1        112       2810.0           0.0          0.0
2        113      10283.0       57483.0          0.0
3        114       5774.0          83.0          0.0
4        115        981.0         912.0         81.0
5        116          0.0        4834.0          4.0
6        117          0.0          98.0         13.0
7        118          0.0           0.0        173.0

In [10]:
df2 = df_revenue.merge(client_base, on='client_id', how='left')
df2

client_id  rzd_revenue  auto_revenue  air_revenue          address
0        111       1093.0           0.0          0.0  Комсомольская 4
1        112       2810.0           0.0          0.0   Энтузиастов 8а
2        113      10283.0       57483.0          0.0  Левобережная 1а
3        114       5774.0          83.0          0.0          Мира 14
4        115        981.0         912.0         81.0        ЗЖБИиДК 1
5        116          0.0        4834.0          4.0    Строителей 18
6        117          0.0          98.0         13.0  Панфиловская 33
7        118          0.0           0.0        173.0     Мастеркова 4